### PR TITLE
Adding RomFS parser

### PR DESCRIPTION
**Pull request by OFFTKP**

### CODE REVIEW

**Comment by OFFTKP** [comment]
@wheremyfoodat give it a sniff before I continue on to file stuff tomorrow

File: src/core/fs/romfs.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        u32* metadataPtr = (u32*)(metadataBase + metadataOffset);
+        metadataPtr += 2;
+        u32 currentDirectoryOffset = *metadataPtr;
+
+        // Loop over all the sibling directories of the first child to get all the children directories
+        while (currentDirectoryOffset != metadataInvalidEntry) {
+            metadataPtr = (u32*)(metadataBase + currentDirectoryOffset);
+            metadataPtr++; // Skip the parent offset
+            u32 siblingDirectoryOffset = *metadataPtr++;
+            metadataPtr++; // Skip the child offset
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:59]: Equivalent to metaDataPtr += 3

File: src/core/fs/romfs.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            metadataPtr++; // Skip the first file offset
+            metadataPtr++; // Skip the next directory in hash table offset
+            u32 nameLength = (*metadataPtr++) / 2;
+
+            // Arbitrary limit
+            if (nameLength > 128) {
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:65]: There is a name limit somewhere in the spec I believe, but eh it's fine for now

File: src/core/fs/ivfc.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
+#include "fs/ivfc.hpp"
+
+namespace IVFC {
+
+    size_t parseIVFC(uintptr_t ivfcStart, IVFC& ivfc) {
+        uintptr_t ivfcPointer = ivfcStart;
+        std::string magicIVFC((char*)ivfcPointer, 4);
```
Comment (wheremyfoodat) [inline src/core/fs/ivfc.cpp:7]: Not a fan, prefer the magic to be checked similar to NCSD

File: src/core/fs/romfs.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
+#include "fs/romfs.hpp"
+#include "fs/ivfc.hpp"
+#include "helpers.hpp"
+#include <cstdio>
+#include <memory>
+#include <map>
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:6]: Why map

File: src/core/fs/romfs.cpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+#include <string>
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:7]: Also needs clang format

File: include/fs/romfs.hpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+#include <string>
+#include <vector>
+#include <memory>
+
+namespace RomFS {
+
+    struct RomFSNode {
+        std::u16string name {};
+        // The file/directory offset relative to the start of the RomFS
+        u64 offset { 0 };
```
Comment (wheremyfoodat) [inline include/fs/romfs.hpp:12]: = 0

File: include/fs/romfs.hpp [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        u64 size { 0 };
+        bool isDirectory { false };
```
Comment (wheremyfoodat) [inline include/fs/romfs.hpp:14]: I understand why this method of construction exists but 99% of the time it is not used for its intended purpose and it looks terrible

File: src/core/fs/romfs.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        printf("%s%s\n", path.c_str(), std::string(node.name.begin(), node.name.end()).c_str());
+        path += std::string(node.name.begin(), node.name.end()) + "/";
+        indentation++;
+        for (auto& directory : node.directories) {
+            printNode(*directory, indentation, path);
+        }
+        indentation--;
+    }
+
+    std::vector<std::unique_ptr<RomFSNode>> parseDirectory(const uintptr_t metadataBase, const uintptr_t metadataOffset) {
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:46]: Const for parameters is an eyesore and mostly pointless

File: src/core/fs/romfs.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            return {};
+        }
+
+        uintptr_t masterHashOffset = RomFS::align(ivfcSize, 0x10);
+        // For a weird reason, the level 3 offset is not the one in the IVFC, instead it's
+        // the first block after the master hash
+        // TODO: Find out why and explain in the comment
+        uintptr_t level3Offset = RomFS::align(masterHashOffset + ivfc.masterHashSize, ivfc.levels[2].blockSize);
+        uintptr_t const level3Base = (uintptr_t)romFS + level3Offset;
+        u32* level3Ptr = (u32*)level3Base;
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:105]: I didn't even know that placement for const is legal
AW4183505_07

File: src/core/fs/romfs.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    };
+
+    inline uintptr_t align(uintptr_t value, uintptr_t alignment) {
+        if (value % alignment == 0)
+            return value;
+
+        return value + (alignment - (value % alignment));
+    }
+
+    inline void printNode(const RomFSNode& node, int indentation, std::string path) {
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:33]: String by copy bad string by const reference good
AW4183505_07

File: src/core/fs/romfs.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        for (int i = 0; i < indentation; i++) {
+            printf("  ");
+        }
+        printf("%s%s\n", path.c_str(), std::string(node.name.begin(), node.name.end()).c_str());
+        path += std::string(node.name.begin(), node.name.end()) + "/";
+        indentation++;
+        for (auto& directory : node.directories) {
+            printNode(*directory, indentation, path);
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:41]: Prefer non recursive solutions wherever possible

File: include/fs/ivfc.hpp [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    struct IVFCLevel {
+        u64 logicalOffset;
+        u64 size;
+        u64 blockSize;
+    };
+
+    struct IVFC {
+        u64 masterHashSize;
+        std::vector<IVFCLevel> levels;
+    };
```
Comment (wheremyfoodat) [inline include/fs/ivfc.hpp:18]: IWYU: File does not include <vector>

**Comment by OFFTKP** [comment]
made the parser non recursive, please review

I'll make the printing thing non recursive too - or remove it, it was only useful to me during debugging

**Comment by wheremyfoodat** [comment]
lptm (looks panda to me)

File: src/core/fs/romfs.cpp [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+		u32 fileDataOffset;
+	};
+
+	inline uintptr_t align(uintptr_t value, uintptr_t alignment) {
+		if (value % alignment == 0) return value;
+
+		return value + (alignment - (value % alignment));
+	}
+
+	inline void printNode(const RomFSNode& node, int indentation) {
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:33]: Why inline here

File: src/core/fs/romfs.cpp [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+		u32 directoryMetadataOffset;
+		u32 directoryMetadataSize;
+		u32 fileHashTableOffset;
+		u32 fileHashTableSize;
+		u32 fileMetadataOffset;
+		u32 fileMetadataSize;
+		u32 fileDataOffset;
+	};
+
+	inline uintptr_t align(uintptr_t value, uintptr_t alignment) {
```
Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:27]: constexpr

Comment (wheremyfoodat) [inline src/core/fs/romfs.cpp:27]: Also this doesn't specify that it's alignUp instead of down

**Review by wheremyfoodat — APPROVED** [review]
Needs to be changed to read the provided ROM block by block properly in the future and not require a fully loaded ROM in memory in the future

**Pull request merged**